core/src/main/java/com/datastax/oss/driver/internal/core/loadbalancing/BasicLoadBalancingPolicy.java

```diff
@@ -325,24 +338,63 @@ protected Queue<Node> maybeAddDcFailover(@Nullable Request request, @NonNull Que
 
           @Override
           protected Object[] computeNodes() {
-            Object[] remoteNodes =
-                liveNodes.dcs().stream()
-                    .filter(Predicates.not(Predicates.equalTo(localDc)))
-                    .flatMap(dc -> liveNodes.dc(dc).stream().limit(maxNodesPerRemoteDc))
-                    .toArray();
-
-            int remoteNodesLength = remoteNodes.length;
-            if (remoteNodesLength == 0) {
-              return EMPTY_NODES;
+            if (!preferredRemoteLocalDcs.isEmpty()) {
+              final List<Object> nodes = new ArrayList<>();
+
+              preferredRemoteLocalDcs.forEach(dc -> addRemoteNodes(nodes, dc));
+
+              liveNodes
+                  .dcs()
+                  .forEach(
+                      dc -> {
+                        if (!preferredRemoteLocalDcs.contains(dc) && !dc.equals(localDc)) {
+                          addRemoteNodes(nodes, dc);
+                        }
+                      });
+
+              if (!nodes.isEmpty()) {
+                return nodes.toArray();
+              } else {
+                return EMPTY_NODES;
+              }
+            } else {
+              Object[] remoteNodes =
+                  liveNodes.dcs().stream()
+                      .filter(Predicates.not(Predicates.equalTo(localDc)))
+                      .flatMap(dc -> liveNodes.dc(dc).stream().limit(maxNodesPerRemoteDc))
+                      .toArray();
+
+              int remoteNodesLength = remoteNodes.length;
+              if (remoteNodesLength == 0) {
+                return EMPTY_NODES;
+              }
+              shuffleHead(remoteNodes, remoteNodesLength);
+              return remoteNodes;
+            }
+          }
+
+          private void addRemoteNodes(final List<Object> nodes, final String dc) {
+            final Object[] remoteNodesPerDc =
+                liveNodes.dc(dc).stream().limit(maxNodesPerRemoteDc).toArray();
+            int remoteNodesLength = remoteNodesPerDc.length;
+            if (remoteNodesLength > 0) {
+              shuffleHeadRemote(remoteNodesPerDc, remoteNodesLength);
+
+              for (int i = 0; i < remoteNodesLength; i++) {
+                nodes.add(remoteNodesPerDc[i]);
+              }
             }
-            shuffleHead(remoteNodes, remoteNodesLength);
-            return remoteNodes;
           }
         };
 
     return new CompositeQueryPlan(local, remote);
   }
 
+  /** Exposed as a protected method so that it can be accessed by tests */
+  protected void shuffleHeadRemote(Object[] currentNodes, int headLength) {
```

Comment:
Not a huge fan of this tbh. I believe you can distinguish calls to `shuffleHead` by inspecting their arguments. See here for an example:

https://github.com/apache/cassandra-java-driver/blob/16260261d3df50fcf24fac1fc2d37896c4a111bf/core/src/test/java/com/datastax/oss/driver/internal/core/loadbalancing/BasicLoadBalancingPolicyDcFailoverTest.java#L67-L70

Reply:
done